cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll

```diff
@@ -14,95 +14,49 @@ private class MallocAllocationFunction extends AllocationFunction {
-        or
-        // CRYPTO_zalloc(size_t num, const char *file, int line)
-        name = "CRYPTO_zalloc" and sizeArg = 0
-        or
-        // CRYPTO_secure_malloc(size_t num, const char *file, int line)
-        name = "CRYPTO_secure_malloc" and sizeArg = 0
-        or
-        // CRYPTO_secure_zalloc(size_t num, const char *file, int line)
-        name = "CRYPTO_secure_zalloc" and sizeArg = 0
-      )
-    )
+    hasGlobalOrStdName("malloc") and // malloc(size)
+    sizeArg = 0
+    or
+    hasGlobalName([
+        "MmAllocateContiguousMemory", // MmAllocateContiguousMemory(size, maxaddress)
+        "MmAllocateContiguousNodeMemory", // MmAllocateContiguousNodeMemory(size, minaddress, maxaddress, bound, flag, prefer)
+        "MmAllocateContiguousMemorySpecifyCache", // MmAllocateContiguousMemorySpecifyCache(size, minaddress, maxaddress, bound, type)
+        "MmAllocateContiguousMemorySpecifyCacheNode", // MmAllocateContiguousMemorySpecifyCacheNode(size, minaddress, maxaddress, bound, type, prefer)
+        "MmAllocateNonCachedMemory", // MmAllocateNonCachedMemory(size)
+        "MmAllocateMappingAddress", // MmAllocateMappingAddress(size, tag)
+        "CoTaskMemAlloc", // CoTaskMemAlloc(size)
+        "kmem_alloc", // kmem_alloc(size, flags)
+        "kmem_zalloc", // kmem_zalloc(size, flags)
+        "CRYPTO_malloc", // CRYPTO_malloc(size_t num, const char *file, int line)
+        "CRYPTO_zalloc", // CRYPTO_zalloc(size_t num, const char *file, int line)
+        "CRYPTO_secure_malloc", // CRYPTO_secure_malloc(size_t num, const char *file, int line)
+        "CRYPTO_secure_zalloc" // CRYPTO_secure_zalloc(size_t num, const char *file, int line)
+      ]) and
+    sizeArg = 0
+    or
+    hasGlobalName([
+        "ExAllocatePool", // ExAllocatePool(type, size)
+        "ExAllocatePoolWithTag", // ExAllocatePool(type, size, tag)
+        "ExAllocatePoolWithTagPriority", // ExAllocatePoolWithTagPriority(type, size, tag, priority)
+        "ExAllocatePoolWithQuota", // ExAllocatePoolWithQuota(type, size)
+        "ExAllocatePoolWithQuotaTag", // ExAllocatePoolWithQuotaTag(type, size, tag)
+        "IoAllocateMdl", // IoAllocateMdl(address, size, flag, flag, irp)
+        "IoAllocateErrorLogEntry", // IoAllocateErrorLogEntry(object, size)
+        "LocalAlloc", // LocalAlloc(flags, size)
+        "GlobalAlloc", // GlobalAlloc(flags, size)
+        "VirtualAlloc" // VirtualAlloc(address, size, type, flag)
+      ]) and
+    sizeArg = 1
+    or
+    hasGlobalName(["HeapAlloc"]) and // HeapAlloc(heap, flags, size)
+    sizeArg = 2
+    or
+    hasGlobalName([
+        "MmAllocatePagesForMdl", // MmAllocatePagesForMdl(minaddress, maxaddress, skip, size)
+        "MmAllocatePagesForMdlEx", // MmAllocatePagesForMdlEx(minaddress, maxaddress, skip, size, type, flags)
+        "MmAllocateNodePagesForMdlEx" // MmAllocateNodePagesForMdlEx(minaddress, maxaddress, skip, size, type, prefer, flags)
+      ]) and
+    sizeArg = 3
```

Comment:
There are a few different ways I could have arranged this and a few similar predicates.  Please comment if you have a preferred design!

Reply:
I think the organization here is good, but given how many environments these classes cover, we should add some comments about where the functions come from.

Reply:
Done.  Bit of a pain tracking the information down but its good to have it.